aircmd/models/base.py

```diff
@@ -45,50 +44,143 @@ class GlobalSettings(BaseSettings):
+ 2 Dependency tracking: In some CI systems, a pipeline run might depend on the successful         
+   completion of its parent or ancestor runs. With the child-knows-parent approach, you can easily
+   check the status of the parent pipelines and determine if the current pipeline run should      
+   proceed or not.                                                                                
+ 3 Debugging and troubleshooting: When a pipeline run fails, you might need to investigate the    
+   cause of the failure by looking at the parent or ancestor runs. The child-knows-parent approach
+   allows you to easily navigate up the pipeline hierarchy to find the relevant information.      
+
+While it's true that some use cases might require traversing the pipeline hierarchy from the root 
+to the leaves, these cases are less common in CI systems. The child-knows-parent approach provides
+more flexibility and ease of use for the typical operations performed on CI pipelines.        
+
+'''
+class Pipeline(BaseModel):
+    name: str
+    parent_pipeline: Optional['Pipeline']
+    dagger_client: dagger.Client
+
+    class Config:
+        arbitrary_types_allowed=True
+
     @classmethod
-    def load_secrets_from_file(cls, secrets_file: str) -> None:
-        """Decrypt the secrets file using SOPS and load it into environment variables."""
-        try:
-            decrypted_secrets = subprocess.check_output(["sops", "-d", secrets_file])
-            with open(".env", "wb") as f:
-                f.write(decrypted_secrets)
-            load_dotenv(".env")
-            os.remove(".env")
-        except subprocess.CalledProcessError as e:
-            raise ValueError(f"Error decrypting secrets file: {e}")
-        except FileNotFoundError:
-            raise ValueError("SOPS command not found. Please install SOPS to use this feature.")
-
-# Mutable. Store pipeline state here. Created at runtime
+    async def create(cls, name: str, parent_pipeline: Optional['Pipeline'] = None) -> 'Pipeline':
+        if parent_pipeline is None:
+            # This is a top-level pipeline. Get the Dagger client from the current PipelineContext.
+            async with PipelineContext() as ctx:
+                dagger_client = ctx.dagger_client
+        else:
+            # This is a nested pipeline. Get the Dagger client from the parent pipeline.
+            dagger_client = parent_pipeline.dagger_client
+
+        # Create the pipeline using the appropriate Dagger client.
+       
+        dagger_client = dagger_client.pipeline(name)
+        return cls(name=name, parent_pipeline=parent_pipeline, dagger_client=dagger_client)
+
+
 class PipelineContext(BaseModel, Singleton):
-    dagger_config: dagger.Config
-    dagger_connection: dagger.Connection
+    _current_level: int = 0
+    logger: logging.Logger
+    current_running_tasks: int = 0
+    max_concurrency: int = 0
+    max_concurrency_per_level: dict[int, int] = {}
```

Comment:
It's ok with immutable types like a tuple. That value is set at definition time. So the problem is if you mutate it in the function, it'll persist for the next function calls because it's a reference to the same thing.